### PR TITLE
W5 - AnhQuach - Thread Handle CountDown Timer

### PR DESCRIPTION
**Pull request by at-anhquach**
#### Related issues (Optional)
Task #66 

#### Description (Required)
- [x] Thread Down Load Image
- [x] CountDown add and remove RecyclerView

### CODE REVIEW

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                });
+            }
+            getActivity().runOnUiThread(new Runnable() {
+                @Override
+                public void run() {
+                    ((ViewImgFragment) ((ThreadActivity) getActivity()).mTabLayoutAdapter.getItem(1)).showPhoto(myBitmap);
+                }
+            });
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Created by anh.quach on 12/19/17.
+ * Render 1 button, 1 ProgressBar
+ */
+public class DownLoadImgFragment extends Fragment {
+    private Button mBtnClick;
+    private int mTotalSize;
+    private int mDownloadedSize = 0;
+    private TextView mTvResutlDownLoad;
+    private ProgressBar mProgressBarDownLoadImg;
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java]: Pls change the name, I don't this is an image.

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                output.write(buffer, 0, bufferLength);
+                myBitmap = BitmapFactory.decodeByteArray(output.toByteArray(), 0, output.size());
+                mDownloadedSize += bufferLength;
+                getActivity().runOnUiThread(new Runnable() {
+                    @SuppressLint("SetTextI18n")
+                    @Override
+                    public void run() {
+                        mProgressBarDownLoadImg.setProgress(mDownloadedSize);
+                        float per = ((float) mDownloadedSize / mTotalSize) * 100;
+                        mTvResutlDownLoad.setText("DownLoad " + mDownloadedSize + "KB/" + mTotalSize + "KB (" + (int) per + "%)");
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java]: Can you define this string to `strings.xml` ?
And using the String format for passing value?

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    public void onBindViewHolder(RecyclerViewViewHolder holder, int position) {
+        holder.onBindData();
+    }
+
+    @Override
+    public int getItemCount() {
+        return mStrs.size();
+    }
+
+    class RecyclerViewViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: Pls make this class to `static` class and define Javadoc.

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_recycler_view_count_down_timer);
+        initViews();
+        mStrs.add("0");
+        mStrs.add("1");
+        mStrs.add("2");
+        initAdapter();
+        new CountDownTimer(180000, 5000) {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: Pls ddefine this number to static variable.

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            int count = 0;
+
+            @Override
+            public void onTick(long longMilisecond) {
+                ++count;
+                if (count % 2 == 0) {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: I don't understand that code, pls explain.

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    private RecyclerView mRecyclerViewCountDownTimer;
+    private List<String> mStrs = new ArrayList<>();
+    private RecyclerViewAdapter mRecyclerViewAdapter;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_recycler_view_count_down_timer);
+        initViews();
+        mStrs.add("0");
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: Why need to add item 0 , 1 , 2

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/ViewImgFragment.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_view_image, container, false);
+        mImgDownLoadedImg = view.findViewById(R.id.imgDownLoaded);
+        return view;
+    }
+
+    public void showPhoto(Bitmap bitmap) {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/ViewImgFragment.java]: Pls remove `public`

File: app/src/at-anhquach/res/layout/fragment_view_image.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <ImageView
```
Comment (at-kietvo) [inline app/src/at-anhquach/res/layout/fragment_view_image.xml]: Pls remove ` LinearLayout`

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    RecyclerViewAdapter(List<String> strs) {
+        mStrs = strs;
+    }
+
+    @Override
+    public RecyclerViewViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        View view = LayoutInflater.from(parent.getContext())
+                .inflate(R.layout.item_countdowntimer, parent, false);
+        return new RecyclerViewViewHolder(view) {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: Is this true ` { } ` ??

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Created by anh.quach on 12/20/17.
+ * Create adapter for 1 rectcler view
+ */
+public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.RecyclerViewViewHolder> {
+    private static List<String> mStrs;
+
+    RecyclerViewAdapter(List<String> strs) {
+        mStrs = strs;
```
Comment (at-dev-support-01) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field vn.asiantech.internship.asynchronous.RecyclerViewAdapter.mStrs from instance method new vn.asiantech.internship.asynchronous.RecyclerViewAdapter(List)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

Comment (at-dev-support-01) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field vn.asiantech.internship.asynchronous.RecyclerViewAdapter.mStrs from instance method new vn.asiantech.internship.asynchronous.RecyclerViewAdapter(List)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            getActivity().runOnUiThread(new Runnable() {
+                public void run() {
+                    mProgressBarDownLoadImg.setMax(mTotalSizeImg);
+                }
+            });
+            byte[] buffer = new byte[1024];
+            int bufferLength;
+            while (0 < (bufferLength = input.read(buffer))) {
+                output.write(buffer, 0, bufferLength);
+                myBitmap = BitmapFactory.decodeByteArray(output.toByteArray(), 0, output.size());
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java]: Pls move Bitmap outside `while` condition

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        mTvResutlDownLoadImg = view.findViewById(R.id.tvDownLoadImg);
+    }
+
+    private void downloadPhoto() {
+        try {
+            URL url = new URL("https://wallpaperbrowse.com/media/images/dd170f37d84b3b21635f2ece8d416afa_large.jpeg");
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setDoInput(true);
+            connection.connect();
+            InputStream input = connection.getInputStream();
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/DownLoadImgFragment.java]: Pls `close()` inputstream, outputstream and connection

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        initAdapter();
+        new CountDownTimer(MILLISINFUTURE, COUNTDOWNINTERVAL) {
+            int countadd = 0;
+            int countdelete = 0;
+
+            @Override
+            public void onTick(long longMilisecond) {
+                ++countadd;
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: change to `mCountAdd`, `mCountDelete`

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    private RecyclerViewAdapter mRecyclerViewAdapter;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_recycler_view_count_down_timer);
+        initViews();
+        mStrs.add("0");
+        initAdapter();
+        new CountDownTimer(MILLISINFUTURE, COUNTDOWNINTERVAL) {
```
Comment (at-kietvo) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: pls refer instance to an variable and close CountDownTimer in `onDestroy()`

File: app/src/at-anhquach/res/layout/activity_thread.xml [hunk 16]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <android.support.design.widget.TabLayout
+        android:id="@+id/tabLayoutThread"
+        android:layout_width="fill_parent"
```
Comment (at-kietvo) [inline app/src/at-anhquach/res/layout/activity_thread.xml]: Pls change to `match_parent`

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java (hunk 14 above)
Comment (at-anhquach) [reply app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: this is local variable, not private

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java (hunk 11 above)
Comment (at-dev-support-01) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field vn.asiantech.internship.asynchronous.RecyclerViewAdapter.mStrs from instance method new vn.asiantech.internship.asynchronous.RecyclerViewAdapter(List)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-anhquach/res/layout/activity_thread.xml (hunk 16 above)
Comment (at-anhquach) [reply app/src/at-anhquach/res/layout/activity_thread.xml]: i changed but this not change

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java (hunk 14 above)
Comment (at-kietvo) [reply app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewCountDownTimerActivity.java]: This variable is not local variable, global variable is a variable define at in 1 Class and outside of 1 method.
As you can see, the variable `countadd` define in `anonymous class` and outside of `onTick` method.

also About convention, the variable must be separate between word, we have `count` and `add` are separate. Pls change it.

File: app/src/at-anhquach/res/layout/activity_thread.xml (hunk 16 above)
Comment (at-kietvo) [reply app/src/at-anhquach/res/layout/activity_thread.xml]: You not yet changed this, pls change `fill_parent` -> `match_parent` and push that code again.

File: app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java (hunk 11 above)
Comment (at-dev-support-01) [inline app/src/at-anhquach/java/vn/asiantech/internship/asynchronous/RecyclerViewAdapter.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field vn.asiantech.internship.asynchronous.RecyclerViewAdapter.mStrs from instance method new vn.asiantech.internship.asynchronous.RecyclerViewAdapter(List)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

**Pull request merged**